### PR TITLE
fix(dom): change EventDelegator bubble to search topElement instead of body

### PR DESCRIPTION
**Pull request by ajmchambers**
- [x] I added new tests for the issue I fixed/built
- [x] I ran `npm test` for the package I'm modifying
- [x] I used `npm run commit` instead of `git commit`

Searching the body will not find the element inside a DocumentFragment or shadowRoot, so the function ends up exiting early and the event is ignored. Added a test for a click inside a DocumentFragment, behaviour is similar to shadowRoot but has better browser support.

ISSUES CLOSED: #412

### CODE REVIEW

**Comment by Widdershin** [comment]
The test is timing out in a bunch of browsers. I wonder why?

**Comment by ajmchambers** [comment]
Hmm yea, I'll see if I can figure out what's going on.

**Comment by staltz** [comment]
Thanks @ajmchambers for this PR. It makes sense. I'll add some comments inline.

File: dom/test/browser/events.js [hunk 1]
```diff
@@ -340,6 +340,52 @@ describe('DOMSource.events()', function () {
...
+      ev.initMouseEvent(
+        `click`,
+        true /* bubble */, true /* cancelable */,
+        window, null,
+        0, 0, 0, 0, /* coordinates */
+        false, false, false, false, /* modifier keys */
+        0 /*left*/, null
+      )
+      el.dispatchEvent(ev)
+    }
```
Comment (staltz) [inline dom/test/browser/events.js]: We use just `element.click()` in other tests, so I think you can remove this helper function.

File: dom/test/browser/events.js [hunk 2]
```diff
@@ -340,6 +340,52 @@ describe('DOMSource.events()', function () {
...
+        assert.strictEqual(ev.target.tagName, 'DIV');
+        assert.strictEqual(ev.target.className, 'clickable');
+        assert.strictEqual(ev.target.textContent, 'Hello');
+        assert.strictEqual(ev.target.parentElement.parentNode instanceof DocumentFragment, true)
+        done();
+      });
+
+    sources.DOM.select(':root').elements().skip(1).take(1).subscribe(root => {
+      const clickable = root.querySelector('.clickable');
+      setTimeout(() => click(clickable));
```
Comment (staltz) [inline dom/test/browser/events.js]: Maybe add more time here, like `setTimeout(() => ...., 80)`.

**Comment by ajmchambers** [comment]
I've been playing around with Chrome 39 (one of the browsers that failed the test) to see if I can figure out whats going on.

It would appear (in Chrome 39 at least) that events aren't bubbling in detached dom.

```javascript
var parent = document.createElement('div')
var child = parent.appendChild(document.createElement('div'));
parent.addEventListener('e', function() {
  console.log('event bubbled to parent');
});
child.addEventListener('e', function() {
  console.log('child dispatched an event');
});
child.dispatchEvent(new CustomEvent('e', { bubbles: true }));
```

Logs both to the console in modern chrome and safari etc, but in chrome 39 it logs the initial child event but not the parent one.

**Comment by mightyiam** [comment]
Congrats on debugging win. Is it out of curiosity or is there a reason to support Chrome 39 specifically?

**Comment by ajmchambers** [comment]
Travis failed because the test I wrote failed in the following browsers:

- iPhone Safari 8.0
- iPhone Safari 9.0
- Chrome 30.0
- Chrome 39.0
- Safari 8.0
- Safari 9.1

So in Chrome 39 (at least) I think my test failed because the event didn't actually bubble - which is the point of my test in the first place. Not sure about the other browsers yet, may be the same issue or something else.

I'm pretty new to this in general, but is there a best practice for tests that you expect to fail in certain situations and not others? (or is that just a bad test)

I suppose you could do this:

```javascript
function eventsBubbleInDetachedDOM() {
  // checks and returns boolean
}

it('should catch bubbling events in a DocumentFragment (and shadowRoot)', function (done) {
  if(eventsBubbleInDetachedDOM()) {
    // run the test
    done();
  } else {
    done();
  }
});
```

or this:

```javascript
function eventsBubbleInDetachedDOM() {
  // checks and returns boolean
}

if(eventsBubbleInDetachedDOM()) {
  it('should catch bubbling events in a DocumentFragment (and shadowRoot)', function (done) {
    // run the test
    done();
  });
}
```

But it doesn't seem right having branching logic for tests.

Any ideas?

(Busy at the moment so won't be able to have a proper look at this again until later on tonight - but when I do get a chance I will make the requested changes and see what happens).

**Comment by staltz** [comment]
You can do something like: https://github.com/cyclejs/cyclejs/blob/master/dom/test/browser/render.js#L213

But I'm curious what exactly don't those old browsers support.

**Comment by ajmchambers** [comment]
Hopefully this fixes it: added a check to see if events will capture or bubble in DocumentFragment and added a test for useCapture. Will check back later. 

Btw should I also be using `npm run commit` when updating a pull request?

**Comment by ajmchambers** [comment]
Ok, now I'm a little confused: dom tests passed ok, but history tests are failing somehow... could my change in the dom EventDelegator have affected history somehow?

Will have a look into it later on. If it was could maybe change my check to something like:

```javascript
if ( 
  !(
    document.body.contains(rawEvent.currentTarget as Node) ||     
    this.topElement.contains(rawEvent.currentTarget as Node)
  )
) {
  return;
} 
```

**Comment by staltz** [comment]
Thanks for the PR!

**Pull request merged**